File: python/2816. Double a Number Represented as a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        reversed = self.reverse(head)
        newHead = ListNode(0)
        pointer = newHead
        carry = 0
        while reversed:
            pointer.next = ListNode((reversed.val * 2 + carry) % 10)
            carry = (reversed.val * 2 + carry) // 10
            reversed = reversed.next
            pointer = pointer.next
        if carry:
            pointer.next = ListNode(carry)
        return self.reverse(newHead.next)
    def reverse(self, head):
        prev = None
        while head:
            head.next, prev, head = prev, head, head.next
        return prev

File: java/2816. Double a Number Represented as a Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode doubleIt(ListNode head) {
        ListNode reversed = reverse(head);
        ListNode newHead = new ListNode();
        ListNode pointer = newHead;
        int carry = 0;
        while (reversed != null) {
            pointer.next = new ListNode((reversed.val * 2 + carry) % 10);
            carry = (reversed.val*2 + carry) / 10;
            reversed = reversed.next;
            pointer = pointer.next;
        }
        if (carry != 0) {
            pointer.next = new ListNode(carry);
        }
        return reverse(newHead.next);
    }
    private ListNode reverse(ListNode head) {
        ListNode previous = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = previous;
            previous = head;
            head = temp;
        }
        return previous;
    }
}